---
output: html_document
---
# Cleaning the Expression Matrix

## Expression QC (UMI) {#exprs-qc}

### Introduction

Once gene expression has been quantified it is summarized as an __expression matrix__ where each row corresponds to a gene (or transcript) and each column corresponds to a single cell. This matrix should be examined to remove poor quality cells which were not detected in either read QC or mapping QC steps. Failure to remove low quality cells at this
stage may add technical noise which has the potential to obscure
the biological signals of interest in the downstream analysis. 

Since there is currently no standard method for performing scRNASeq the expected values for the various QC measures that will be presented here can vary substantially from experiment to experiment. Thus, to perform QC we will be looking for cells which are outliers with respect to the rest of the dataset rather than comparing to independent quality standards. Consequently, care should be taken when comparing quality metrics across datasets collected using different protocols.


### Tung dataset

To illustrate cell QC, we consider a
[dataset](http://jdblischak.github.io/singleCellSeq/analysis/) of
 induced pluripotent stem cells generated from three different individuals [@Tung2017-ba] in [Yoav Gilad](http://giladlab.uchicago.edu/)'s lab at the
University of Chicago. The experiments were carried out on the
Fluidigm C1 platform and to facilitate the quantification both unique
molecular identifiers (UMIs) and ERCC _spike-ins_ were used. The data files are located in the `tung` folder in your working directory. These files are the copies of the original files made on the 15/03/16. We will use these copies for reproducibility purposes.

```{r exprs-qc0, echo=FALSE}
library(knitr)
opts_chunk$set(cache = TRUE, out.width='90%', fig.align = 'center')
```

```{r exprs-qc1, message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(scater)
library(EnsDb.Hsapiens.v86)
options(stringsAsFactors = FALSE)
```

```{r exprs-qc2}
molecules <- read.delim("data/tung/molecules.txt",row.names=1)
annotation <- read.delim("data/tung/annotation.txt")
```

```{r exprs-qc3}
head(molecules[ , 1:3])
head(annotation)
```

```{r exprs-qc4}
## set altExp to contain ERCC, remove ERCC from the main object
umi <- SingleCellExperiment(assays = list(counts = as.matrix(molecules)), colData = annotation)
altExp(umi,"ERCC") <- umi[grep("^ERCC-",rownames(umi)), ]
umi <- umi[grep("^ERCC-",rownames(umi),invert = T), ]
```

```{r exprs-qc5}
## define is.mito vector (T/F values, length same as number of genes in the main assay)
g = genes(EnsDb.Hsapiens.v86)
MT_names = g[seqnames(g)=="MT"]$gene_id
is.mito = rownames(umi) %in% MT_names
```

```{r exprs-qc6}
## set all PerCell and PerFeature QC metrics
umi_cell <- perCellQCMetrics(umi,subsets=list(Mito=is.mito))
umi_feature <- perFeatureQCMetrics(umi)
head(umi_cell)
head(umi_feature)
umi <- addPerCellQC(umi, subsets=list(Mito=is.mito))
umi <- addPerFeatureQC(umi)
```

```{r exprs-qc7}
## filtering options: manual 
hist(
    umi$total,
    breaks = 100
)
abline(v = 25000, col = "red")
```

```{r exprs-qc8}
hist(
  umi_cell$detected,
  breaks = 100
)
abline(v = 7000, col = "red")
```

```{r exprs-qc9}
## ADAPTIVE threshold - filters whatever is more than 3 standard deviations away from the mean
qc.lib2 <- isOutlier(umi_cell$sum, log=TRUE, type="lower")
attr(qc.lib2, "thresholds")
qc.nexprs2 <- isOutlier(umi_cell$detected, log=TRUE, type="lower")
attr(qc.nexprs2, "thresholds")
qc.spike2 <- isOutlier(umi_cell$altexps_ERCC_percent, type="higher")
attr(qc.spike2, "thresholds")
qc.mito2 <- isOutlier(umi_cell$subsets_Mito_percent, type="higher")
attr(qc.mito2, "thresholds")
discard2 <- qc.lib2 | qc.nexprs2 | qc.spike2 | qc.mito2
DataFrame(LibSize=sum(qc.lib2), NExprs=sum(qc.nexprs2), SpikeProp=sum(qc.spike2), MitoProp=sum(qc.mito2), Total=sum(discard2))
```

```{r exprs-qc10}
## ADAPTIVE threshold in one command
reasons <- quickPerCellQC(umi_cell, sub.fields=c("subsets_Mito_percent", "altexps_ERCC_percent"))
colSums(as.matrix(reasons))
```

```{r exprs-qc11}
## Add discarded data as own coldata assay called discard
umi$discard <- reasons$discard
```

```{r exprs-qc12}
## Plotting various coldata assays against each other
plotColData(umi, x="sum", y="subsets_Mito_percent", colour_by="discard")
plotColData(umi, x="sum", y="detected", colour_by="discard")
plotColData(umi, x="altexps_ERCC_percent", y="subsets_Mito_percent",colour_by="discard")
```

```{r exprs-qc13}
## Plot coldata with splitting by batches to see if there are substantial differences
plotColData(umi, x="sum", y="detected", colour_by="discard", other_fields = "individual") + facet_wrap(~individual)
plotColData(umi, x="sum", y="detected", colour_by="discard", other_fields = "replicate") + facet_wrap(~replicate)
```

```{r exprs-qc14}
## Plotting highest expressed genes
plotHighestExprs(umi, exprs_values = "counts")
```

```{r exprs-qc15}
## Keeping highly expressed genes
keep_feature <- nexprs(umi,byrow = TRUE,detection_limit = 1) >= 2
rowData(umi)$use <- keep_feature
table(keep_feature)
``` 
 
```{r exprs-qc16}
assay(umi, "logcounts_raw") <- log2(counts(umi) + 1)
reducedDim(umi) <- NULL
```

Save the data:
```{r exprs-qc17}
saveRDS(umi, file = "data/tung/umi.rds")
```

### Big Exercise

Perform exactly the same QC analysis with read counts of the same Blischak data. Use `tung/reads.txt` file to load the reads. Once you have finished please compare your results to ours (next chapter).

### sessionInfo()

```{r exprs-qc18, echo=FALSE}
sessionInfo()
```
