---
output: html_document
---

```{r seurat0, echo=FALSE}
library(knitr)
opts_chunk$set(cache = TRUE, fig.align = "center")
```

```{r seurat1, echo=TRUE, message=FALSE, warning=FALSE}
set.seed(1234567)
```

# Seurat {#seurat-chapter}

[Seurat](http://satijalab.org/seurat/) was originally developed as a clustering tool for scRNA-seq data, however in the last few years the focus of the package has become less specific and at the moment `Seurat` is a popular R package that can perform QC, analysis, and exploration of scRNA-seq data, i.e. many of the tasks covered in this course.

__Note__ We recommend using `Seurat` for datasets with more than $5000$ cells. For smaller dataset a good alternative will be `SC3`.

__Note__ In this chapter we use an exact copy of [this tutorial](https://satijalab.org/seurat/pbmc3k_tutorial.html).

## Setup the Seurat Object

We will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found [here](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).

We start by reading in the data. All features in Seurat have been configured to work with sparse matrices which results in significant memory and speed savings for Drop-seq/inDrop/10x data.

```{r seurat2}
library(Seurat)
library(dplyr)
library(cowplot)

pbmc.data <- Read10X(data.dir = "data/pbmc3k_filtered_gene_bc_matrices/hg19/")

dense.size <- object.size(x = as.matrix(x = pbmc.data))
dense.size

sparse.size <- object.size(x = pbmc.data)
sparse.size

dense.size/sparse.size

pbmc <- CreateSeuratObject(pbmc.data, min.cells = 3, min.features  = 200, project = "10X_PBMC")
```

```{r seurat3}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r seurat4}
FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

```{r seurat5}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

```{r seurat6}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{r seurat7}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst",  nfeatures = 2000)
```

To view the output of the FindVariableFeatures output we use this function. The genes appear not to be stored in the object, but can be accessed this way.
```{r seurat8}
top10 <- head(VariableFeatures(pbmc), 10)
top10
```

```{r seurat9}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

```{r seurat10}
pbmc <- RunPCA(pbmc,  features = VariableFeatures(object = pbmc), verbose = FALSE)
```

```{r seurat11}
DimPlot(pbmc, reduction = "pca")
```

```{r seurat12}
FeaturePlot(pbmc, features = "percent.mt")
```

```{r seurat13}
FeatureScatter(pbmc, feature1 = "MS4A1", feature2 = "PC_1")
FeatureScatter(pbmc, feature1 = "MS4A1", feature2 = "CD3D")
```

```{r seurat14}
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

```{r seurat15}
VlnPlot(pbmc, features = c("LYZ", "CCL5", "IL32"))
```

```{r seurat16}
# Heatmaps
DimHeatmap(pbmc, dims = 1:6, cells = 1000, balanced = T)
```

```{r seurat17}
ElbowPlot(pbmc)
```

```{r seurat18}
pbmc <- FindNeighbors(pbmc, dims = 1:20)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

```{r seurat19}
pbmc <- RunUMAP(pbmc, dims = 1:20)
DimPlot(pbmc)
```

```{r seurat20}
saveRDS(pbmc, file = "data/pbmc_tutorial.rds")
```

```{r seurat21}
cluster1.markers <- FindMarkers(object = pbmc, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)
```

```{r seurat22}
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)
```

```{r seurat23}
library(dplyr)
pbmc.markers <- FindAllMarkers(pbmc, only.pos = T, min.pct = 0.25, logfc.threshold = 0.25)
dim(pbmc.markers)
top3_markers <- as.data.frame(pbmc.markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC))
top3_markers
```

```{r seurat24}
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, thresh.use = 0.25, test.use = "roc", only.pos = TRUE)
```

```{r seurat25}
VlnPlot(pbmc, features =c("NKG7", "PF4"))
```

```{r seurat26}
FeaturePlot(object = pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))
```

```{r seurat27}
saveRDS(pbmc, file = "data/pbmc3k_final.rds")
```

```{r seurat28, echo=FALSE}
sessionInfo()
```
